# Applying Advanced Transformations (Core)
- Zach Hanson

## Libraries

In [1]:
#Standard
import pandas as pd
import numpy as np

#JSON
import json

## Load in Data

In [2]:
info_df = pd.read_csv('Data/superhero_info.csv')
info_df.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [3]:
powers_df = pd.read_csv('Data/superhero_powers.csv')
powers_df.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

## Split "Hero|Publisher" Column

In [4]:
info_df[['Hero', 'Publisher']] = info_df['Hero|Publisher'].str.split('|', expand = True)
info_df.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

- Looks like the split was successful
    - Can drop "Hero|Publisher" Column now

In [5]:
info_df = info_df.drop(columns = ['Hero|Publisher'])
info_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

## Split Height and Weight Measurements

In [6]:
#Looking at single measurement value
measurement = info_df.loc[0, 'Measurements']
print(type(measurement))
measurement

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [7]:
measurement = measurement.replace("'", '"')
measurement

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [8]:
fixed_meas = json.loads(measurement)
print(type(fixed_meas))

<class 'dict'>


In [9]:
info_df['Measurements'] = info_df['Measurements'].str.replace("'",'"')
## Apply the json.loads to the full column
info_df['Measurements'] = info_df['Measurements'].apply(json.loads)
info_df['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [10]:
test_meas = info_df.loc[0, 'Measurements']
print(type(test_meas))
test_meas

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [11]:
height_weight = info_df['Measurements'].apply(pd.Series)
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [12]:
#Splitting height column so we can change to numerical values
height_weight[['Height', 'units_height']] = height_weight['Height'].str.split(' ', expand = True)
height_weight

Height    Weight units_height
0    203.0  441.0 kg           cm
1    191.0   65.0 kg           cm
2    185.0   90.0 kg           cm
3    203.0  441.0 kg           cm
4    193.0  122.0 kg           cm
..     ...       ...          ...
458  183.0   83.0 kg           cm
459  165.0   52.0 kg           cm
460   66.0   17.0 kg           cm
461  170.0   57.0 kg           cm
462  185.0   81.0 kg           cm

[463 rows x 3 columns]

In [13]:
#Splitting weight column so we can change to numerical values
height_weight[['Weight', 'units_weight']] = height_weight['Weight'].str.split(' ', expand = True)
height_weight

Height Weight units_height units_weight
0    203.0  441.0           cm           kg
1    191.0   65.0           cm           kg
2    185.0   90.0           cm           kg
3    203.0  441.0           cm           kg
4    193.0  122.0           cm           kg
..     ...    ...          ...          ...
458  183.0   83.0           cm           kg
459  165.0   52.0           cm           kg
460   66.0   17.0           cm           kg
461  170.0   57.0           cm           kg
462  185.0   81.0           cm           kg

[463 rows x 4 columns]

In [14]:
#Getting unique values for height to make sure theyre all the same units
height_weight['units_height'].value_counts()

cm    463
Name: units_height, dtype: int64

- All in cm

In [15]:
#Getting unique values for weight to make sure theyre all the same units
height_weight['units_weight'].value_counts()

kg    463
Name: units_weight, dtype: int64

- All in kg

In [16]:
#Adding units to column names
height_weight = height_weight.rename(columns={'Height': 'Height (cm)',
                                             'Weight': 'Weight (kg)'})
height_weight

Height (cm) Weight (kg) units_height units_weight
0         203.0       441.0           cm           kg
1         191.0        65.0           cm           kg
2         185.0        90.0           cm           kg
3         203.0       441.0           cm           kg
4         193.0       122.0           cm           kg
..          ...         ...          ...          ...
458       183.0        83.0           cm           kg
459       165.0        52.0           cm           kg
460        66.0        17.0           cm           kg
461       170.0        57.0           cm           kg
462       185.0        81.0           cm           kg

[463 rows x 4 columns]

In [17]:
#Removing the units columns
height_weight = height_weight.drop(columns=['units_height', 'units_weight'])
height_weight

Height (cm) Weight (kg)
0         203.0       441.0
1         191.0        65.0
2         185.0        90.0
3         203.0       441.0
4         193.0       122.0
..          ...         ...
458       183.0        83.0
459       165.0        52.0
460        66.0        17.0
461       170.0        57.0
462       185.0        81.0

[463 rows x 2 columns]

In [18]:
#Adding to main dataframe
info_df = pd.concat((info_df, height_weight), axis = 1)
info_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher Height (cm) Weight (kg)  
0      Marvel Comics       203.0       441.0  
1  Dark Horse Comics       191.0        65.0  
2          DC Comics       185.0        90.0  
3      Marvel Comics       203.0       441.0  
4      Marvel Comics       193.0       122.0

In [19]:
#Dropping Measurements column as its redundant now
info_df = info_df.drop(columns=['Measurements'])
info_df.head(3)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   
2   Male        Ungaran      good    No Hair      blue        red    Abin Sur   

           Publisher Height (cm) Weight (kg)  
0      Marvel Comics       203.0       441.0  
1  Dark Horse Comics       191.0        65.0  
2          DC Comics       185.0        90.0

In [20]:
info_df = info_df.astype({'Height (cm)': 'float'})
info_df = info_df.astype({'Weight (kg)': 'float'})

In [21]:
info_df.dtypes

Gender          object
Race            object
Alignment       object
Hair color      object
Eye color       object
Skin color      object
Hero            object
Publisher       object
Height (cm)    float64
Weight (kg)    float64
dtype: object

- Height and Weight are now separate columns and numerical values

In [38]:
info_df.shape

(463, 10)

## OHE Powers Column

In [22]:
test = powers_df.loc[0, 'Powers']
print(type(test))

<class 'str'>


In [23]:
test

'Agility,Super Strength,Stamina,Super Speed'

- Need to convert string to list

In [24]:
powers_df['powers_split'] = powers_df['Powers'].str.split(",")

In [25]:
test = powers_df.loc[0, 'powers_split']
print(type(test))

<class 'list'>


- Successfully changed to list

In [26]:
powers_df.loc[5, 'powers_split']

['Dimensional Awareness',
 'Flight',
 'Intelligence',
 'Super Strength',
 'Size Changing',
 'Super Speed',
 'Teleportation',
 'Magic',
 'Dimensional Travel',
 'Immortality',
 'Invulnerability',
 'Molecular Manipulation',
 'Energy Manipulation',
 'Power Cosmic']

In [27]:
powers_df['powers_split'].value_counts()

[Intelligence]                                                                                                                                                                                                                                                                          8
[Durability, Super Strength]                                                                                                                                                                                                                                                            5
[Agility, Stealth, Marksmanship, Weapons Master, Stamina]                                                                                                                                                                                                                               4
[Marksmanship]                                                                                                                                            

In [28]:
exploded = powers_df.explode('powers_split')
exploded.head(10)

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

          powers_split  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing  
1           Durability  
1            Longevity  
1       Super Strength  
1              Stamina  
1           Camouflage

In [29]:
#Saving unique values from exploded column
cols_to_make = exploded['powers_split'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [31]:
for col in cols_to_make:
    powers_df[col] = powers_df['Powers'].str.contains(col)
    
powers_df.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        powers_split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

- Have our one hot encoded powers now, can delete original column

In [33]:
powers_df = powers_df.drop(columns=['Powers', 'powers_split'])
powers_df.head()

hero_names  Agility  Super Strength  Stamina  Super Speed  \
0      3-D Man     True            True     True         True   
1       A-Bomb    False            True     True        False   
2   Abe Sapien     True            True     True        False   
3     Abin Sur    False           False    False        False   
4  Abomination    False            True     True         True   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   
2                 True        True       True       False            False   
3                False       False      False       False            False   
4                 True       False      False       False            False   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   
2  ...            False        False       False              False   
3  ...            False        False       False              False   
4  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   
2       False       False          False               False    False   
3       False       False          False               False    False   
4       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  
2           False  
3           False  
4           False  

[5 rows x 168 columns]

In [37]:
powers_df.shape

(667, 168)

## Joining two Dataframes

In [34]:
merged_df = pd.merge(info_df, powers_df, left_on='Hero', right_on='hero_names')
merged_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height (cm)  Weight (kg)  ...  \
0         A-Bomb      Marvel Comics        203.0        441.0  ...   
1     Abe Sapien  Dark Horse Comics        191.0         65.0  ...   
2       Abin Sur          DC Comics        185.0         90.0  ...   
3    Abomination      Marvel Comics        203.0        441.0  ...   
4  Absorbing Man      Marvel Comics        193.0        122.0  ...   

  Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0           False        False       False              False       False   
1           False        False       False              False       False   
2           False        False       False              False       False   
3           False        False       False              False       False   
4           False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 178 columns]

In [36]:
merged_df.shape

(463, 178)

- There was only 463 rows in the "superhero_info" dataframe, so having 463 rows after merging seems correct despite having 667 rows for the "superhero_powers" dataframe

## Questions

### Weight of Super Speed vs. Non Super Speed

In [42]:
merged_df.groupby('Super Speed')['Weight (kg)'].mean().round(2)

Super Speed
False    101.77
True     129.40
Name: Weight (kg), dtype: float64

### Average Height for each Publisher

In [40]:
merged_df.groupby('Publisher')['Height (cm)'].mean().round(2)

Publisher
DC Comics            181.92
Dark Horse Comics    176.91
George Lucas         159.60
Image Comics         211.00
Marvel Comics        191.55
Shueisha             171.50
Star Trek            181.50
Team Epic TV         180.75
Unknown              178.00
Name: Height (cm), dtype: float64